Update unlinking a Slack app

robincornett opened the pull request:
Fixes #131 

Proposed changes:
1. Send the oAuth token as an `Authorization` header instead of appending to the URL.
2. Use `GET` instead of `POST` for revoking.
3. Restrict revoking to administrator level users, and update the logic in that function to be a little more readable.
4. When viewing the settings screen for the extension, send a quick authorization check to Slack to see if the token is still valid. If it's not, delete the setting.
5. Rework the logic for setting the oAuth token--a bit more verbose but easier to read for me.

mihaijoldis commented:
While testing this I was able to constantly make it brek the connection to the App. Sometimes after several link/unlin attempts and navigating to a non EDD page so that the URL query params are reset the authentication process just reloads the EDD slack settings page.

Something inside the store_oauth_token() method is causing this as the code returns early here
```
		if ( 'saving' !== $state || ! $token_type ) {
			return false;
		}
```
The `$_GET` dump right above that when it fails looks like this
```
[12-May-2021 14:51:59 UTC] Start GET: Array
(
    [post_type] => download
    [page] => edd-settings
    [tab] => extensions
    [section] => edd-slack-settings
    [code] => 2056589097684.2057878671154.9cff8cb9ed31269c840cfcdb198fea5ffd0aa3c6cdf57fafb787b9624b4e8232
    [state] => 

)
```

robincornett commented:
@mihaijoldis I've updated the linking behavior to add a redirect after checking in with slack so that the token is not...exposed, I guess, in the URL. Also added more checks to the token setting method. Can you check again and see if you can still break it? (Seriously I cannot.)

mihaijoldis commented:
@robincornett managed to break on the second attempt to reconnect.
added the debug entries I had before and again it seems to return `if ( 'saving' !== $state || ! $token_type ) {`

mihaijoldis commented:
Here is a small recording: https://www.loom.com/share/873c8597612743bfb8b3a938b5192de2
You can see that one the modal window closes the main settings page just reloads.

robincornett commented:
@mihaijoldis sorry, can you check again? I just tried it on the testing site and it linked successfully for me. I adjusted the behavior so that the user should not see the code in the URL again--once it's been recorded, the site redirects to `https://edd.thepatternmall.com/wp-admin/edit.php?post_type=download&page=edd-settings&tab=extensions&section=edd-slack-settings&edd_slack_oauth=success&edd_slack_type=updated`.

mihaijoldis commented:
@robincornett its deffinetly better now as I can constantly get it to connect fine. In a few tests I ran like 10-15 quick connect/disconnects it only broke once. The URL on the EDD it redirected to was

`https://edd.thepatternmall.com/wp-admin/edit.php?post_type=download&page=edd-settings&tab=extensions&section=edd-slack-settings&code=2056589097684.2088391303457.95719a88aced88a1e0aea594f2871ec9e4a4dfd0f4a18e267c5ea2acc5d23104&state`

ashleyfae commented on core/ssl-only/class-edd-slack-app-oauth-settings.php:
```diff
@@ -269,148 +319,214 @@ public function add_slack_invites_oauth_register( $args ) {
+					$redirect_uri
+				);
+
+				wp_safe_redirect( $redirect_uri );
```
```suggestion
				wp_safe_redirect( $redirect_uri );
				exit;
```

Without the exit, it actually keeps executing the rest of the code and you end up on the success page.

ashleyfae commented on core/ssl-only/class-edd-slack-app-oauth-settings.php:
```diff
@@ -269,148 +319,214 @@ public function add_slack_invites_oauth_register( $args ) {
+					$redirect_uri
+				);
+			}
+			wp_safe_redirect( $redirect_uri );
```
```suggestion
			wp_safe_redirect( $redirect_uri );
			exit;
```

ashleyfae commented on core/ssl-only/class-edd-slack-app-oauth-settings.php:
```diff
@@ -187,52 +194,97 @@ public function add_oauth_registration_button( $args ) {
+					'https://slack.com/oauth/authorize'
+				);
+				?>
+				<a href="<?php echo $slack_uri; ?>" target="_self" class="edd-slack-app-auth button button-primary" data-token_type="main">
```
```suggestion
				<a href="<?php echo esc_url( $slack_uri ); ?>" target="_self" class="edd-slack-app-auth button button-primary" data-token_type="main">
```

ashleyfae reviewed (changes requested):
This is working for me! Just requested a couple improvements, then we'll be good to go.
